### ISSUE TITLE
Feature - Add GPG controls when the binaries are downloaded to enchance security

### ISSUE THREAD

**Issue by mat1e** [eirslett/frontend-maven-plugin#900]
To enchance the security of the plugin, I want to add the GPG control when a binary package like Yarn, Node, Npm... is downloaded.

We need to add multiple actions before downloading each binary : 

- Download of the shasum file.
- Check shasum file GPG signature.
- Get the hash of matching binary in the shasum file.
- Check the binary hash.
- If ok, copy the binary in the project.

**Comment by eirslett** [comment]
It wouldn't really work if you download the shasum file as well. If some attacker want to create a spoofed node.js binary and host it, they could also spoof the shasum file, and go undetected.

**Comment by mat1e** [comment]
@eirslett, we can verify the integrity of the shasum file like described in the [the README of node](https://github.com/nodejs/node#verifying-binaries).

**Comment by bmhm** [comment]
@eirslett do you allow PRs? I agree that a digest check from the same source would not be too helpful, but a signature check (including verifying to a known fingerprint) would be more useful. The question remains: Do we trust this plugin to maintain the fingerprint? Or can users override the fingerprint? ;-)

**Comment by eirslett** [comment]
To me, it just sounds like a lot of extra complexity for the added security value. One option is to make a separate Maven plugin that does the verifying?

**Comment by bmhm** [comment]
Yes, there is the maven-gpg-plugin. [The maven-dependency-plugin does not check its dependencies itself](https://github.com/apache/maven-dependency-plugin/blob/master/src/main/java/org/apache/maven/plugins/dependency/GetMojo.java).
Anyway, if there is a gpg(2) or pgp binary, that would be realy easy to do. Also, the default mode could be "verify-if-gpg-present" and maybe a forced mode (which fails if no gpg binary is present) and a lax mode (which will only print warnings).

And yes, this could be put into a separate plugin, as long as there is a suitable API.

If you'd want such a thing, maybe we would want to create a github organization and move all stuff over there to have a common namespace.

**Comment by eirslett** [comment]
You could try using BouncyCastle for GPG verification, then everything is implemented in Java and you don't need a separate binary (or so I think). A separate organization isn't really needed; if you want to write such a plugin, you could start on your personal account first, and move it later on if there is a need. I'd be happy to provide a link in the README to such a plugin.

**Comment by mat1e** [comment]
@bmhm I'm available to work on it if needed.

**Comment by bmhm** [comment]
@mat1e I do not need it atm. I’d rather need my PR merged: https://github.com/eirslett/frontend-maven-plugin/pull/927
But thanks for asking.
